File: JobTracker/Services/MboxParser.swift
//
//  MboxParser.swift
//  JobTracker
//
//  Streaming parser for Google Takeout mbox archives. Reads the file in
//  chunks (never loads the whole archive into memory), splits on mbox
//  "From " separators, decodes MIME (RFC 2047 headers, quoted-printable /
//  base64 bodies, multipart), and yields FetchedMessage values compatible
//  with the rest of the pipeline.
//
//  Takeout specifics used here:
//  - Each message starts with "From <decimal-gm-thrid>@xxx <date>"
//  - "X-GM-THRID" (decimal) converts to the hex threadId the Gmail API uses
//  - "X-Gmail-Labels" lets us skip Spam / Trash outright
//

import Foundation

enum MboxParserError: LocalizedError {
    case cannotOpen(String)

    var errorDescription: String? {
        switch self {
        case .cannotOpen(let path): return "Cannot open mbox file: \(path)"
        }
    }
}

nonisolated struct MboxParseSummary {
    var totalMessages = 0
    var skippedSpamTrash = 0
    var candidates = 0
}

/// `nonisolated`: parsing runs in a detached task off the main actor.
nonisolated enum MboxParser {

    /// Streams the mbox at `url`, returning only messages that pass the
    /// job-related prefilter (and are not Spam/Trash). `progress` receives
    /// a 0...1 fraction of file bytes consumed.
    static func collectJobCandidates(
        url: URL,
        progress: ((Double) -> Void)? = nil
    ) throws -> (messages: [FetchedMessage], summary: MboxParseSummary) {

        guard let handle = try? FileHandle(forReadingFrom: url) else {
            throw MboxParserError.cannotOpen(url.path)
        }
        defer { try? handle.close() }

        let fileSize = (try? FileManager.default.attributesOfItem(atPath: url.path)[.size] as? Int) ?? 0
        let chunkSize = 4 * 1024 * 1024
        let separator = Data("\nFrom ".utf8)

        var buffer = Data()
        var bytesConsumed = 0
        var summary = MboxParseSummary()
        var results: [FetchedMessage] = []

        func handleMessage(_ raw: Data) {
            summary.totalMessages += 1
            guard let message = parseMessage(raw) else { return }
            if message.labels.contains("Spam") || message.labels.contains("Trash") {
                summary.skippedSpamTrash += 1
                return
            }
            guard HeuristicPrefilter.isLikelyJobRelated(
                sender: message.fetched.sender,
                subject: message.fetched.subject,
                body: message.fetched.body,
                isOutgoing: message.fetched.isOutgoing
            ) else { return }
            summary.candidates += 1
            results.append(message.fetched)
        }

        while true {
            guard let chunk = try? handle.read(upToCount: chunkSize), !chunk.isEmpty else { break }
            buffer.append(chunk)
            bytesConsumed += chunk.count
            if fileSize > 0 { progress?(Double(bytesConsumed) / Double(fileSize)) }

            // Extract complete messages; keep the trailing partial one buffered.
            var searchStart = buffer.startIndex
            while let range = buffer.range(of: separator, in: searchStart..<buffer.endIndex) {
                // Validate this is a real mbox boundary: "From <digits>@xxx"
                guard isTakeoutBoundary(buffer, atSeparator: range) else {
                    searchStart = range.upperBound
                    continue
                }
                let messageData = buffer.subdata(in: buffer.startIndex..<range.lowerBound)
                if !messageData.isEmpty { handleMessage(messageData) }
                // +1 keeps the "From " (drops only the "\n").
                buffer.removeSubrange(buffer.startIndex..<(range.lowerBound + 1))
                searchStart = buffer.startIndex
            }
        }

        // Final message.
        if !buffer.isEmpty { handleMessage(buffer) }

        return (results, summary)
    }

    /// Checks that the bytes after "\nFrom " look like "<digits>@xxx ".
    private static func isTakeoutBoundary(_ data: Data, atSeparator range: Range<Data.Index>) -> Bool {
        var index = range.upperBound
        var sawDigit = false
        while index < data.endIndex, data[index] >= UInt8(ascii: "0"), data[index] <= UInt8(ascii: "9") {
            sawDigit = true
            index += 1
        }
        guard sawDigit, index < data.endIndex, data[index] == UInt8(ascii: "@") else { return false }
        return true
    }

    // MARK: - Single message

    private struct ParsedMessage {
        let fetched: FetchedMessage
        let labels: Set<String>
    }

    private static func parseMessage(_ raw: Data) -> ParsedMessage? {
        // Split headers / body at the first blank line.
        let crlfSep = Data("\r\n\r\n".utf8)
        let lfSep = Data("\n\n".utf8)
        var headerData = raw
        var bodyData = Data()
        if let r = raw.range(of: crlfSep) ?? raw.range(of: lfSep) {
            headerData = raw.subdata(in: raw.startIndex..<r.lowerBound)
            bodyData = raw.subdata(in: r.upperBound..<raw.endIndex)
        }

        guard let headerText = String(data: headerData, encoding: .utf8)
            ?? String(data: headerData, encoding: .isoLatin1) else { return nil }

        let headers = unfoldHeaders(headerText)
        func header(_ name: String) -> String { headers[name.lowercased()] ?? "" }

        let labels = Set(header("X-Gmail-Labels").components(separatedBy: ",")
            .map { $0.trimmingCharacters(in: .whitespaces) })

        let threadIdHex: String = {
            if let decimal = UInt64(header("X-GM-THRID")) { return String(decimal, radix: 16) }
            return ""
        }()

        let messageId = header("Message-ID")
            .trimmingCharacters(in: CharacterSet(charactersIn: "<> \t"))

        let subject = decodeRFC2047(header("Subject"))
        let sender = decodeRFC2047(header("From"))
        let date = parseDate(header("Date")) ?? Date.distantPast

        let body = extractBody(
            bodyData,
            contentType: header("Content-Type"),
            transferEncoding: header("Content-Transfer-Encoding"),
            depth: 0
        )
        let cleanBody = String(body.prefix(8000))

        let fetched = FetchedMessage(
            id: messageId.isEmpty ? "mbox-\(threadIdHex)-\(raw.count)" : messageId,
            threadId: threadIdHex,
            sender: sender,
            subject: subject,
            snippet: String(cleanBody.replacingOccurrences(of: "\n", with: " ").prefix(160)),
            body: cleanBody,
            isOutgoing: labels.contains("Sent"),
            date: date
        )
        return ParsedMessage(fetched: fetched, labels: labels)
    }

    // MARK: - Headers

    /// Unfolds continuation lines and returns a [lowercased-name: value] map
    /// (first occurrence wins).
    private static func unfoldHeaders(_ text: String) -> [String: String] {
        var result: [String: String] = [:]
        var currentName: String?
        var currentValue = ""

        func commit() {
            if let name = currentName, result[name] == nil {
                result[name] = currentValue.trimmingCharacters(in: .whitespaces)
            }
        }

        for line in text.replacingOccurrences(of: "\r\n", with: "\n").components(separatedBy: "\n") {
            if line.hasPrefix(" ") || line.hasPrefix("\t") {
                currentValue += " " + line.trimmingCharacters(in: .whitespaces)
            } else if let colon = line.firstIndex(of: ":") {
                commit()
                currentName = String(line[..<colon]).lowercased()
                currentValue = String(line[line.index(after: colon)...])
            }
        }
        commit()
        return result
    }

    /// Decodes RFC 2047 encoded words: =?charset?B|Q?data?=
    static func decodeRFC2047(_ value: String) -> String {
        guard value.contains("=?") else { return value }
        let pattern = "=\\?([^?]+)\\?([BbQq])\\?([^?]*)\\?="
        guard let regex = try? NSRegularExpression(pattern: pattern) else { return value }

        var result = value
        // Iterate until no encoded words remain (handles adjacent words).
        var iterations = 0
        while iterations < 20,
              let match = regex.firstMatch(in: result, range: NSRange(result.startIndex..., in: result)),
              let whole = Range(match.range, in: result),
              let charsetRange = Range(match.range(at: 1), in: result),
              let encRange = Range(match.range(at: 2), in: result),
              let dataRange = Range(match.range(at: 3), in: result) {
            iterations += 1
            let charset = String(result[charsetRange]).lowercased()
            let encoding = String(result[encRange]).uppercased()
            let payload = String(result[dataRange])

            let decodedData: Data?
            if encoding == "B" {
                decodedData = Data(base64Encoded: payload)
            } else {
                let qp = payload.replacingOccurrences(of: "_", with: " ")
                decodedData = decodeQuotedPrintable(qp)
            }

            let text: String
            if let data = decodedData {
                let enc: String.Encoding = charset.contains("8859") ? .isoLatin1 : .utf8
                text = String(data: data, encoding: enc)
                    ?? String(data: data, encoding: .isoLatin1) ?? ""
            } else {
                text = ""
            }
            result.replaceSubrange(whole, with: text)
        }
        return result
    }

    // MARK: - Date

    private static let dateFormats = [
        "EEE, d MMM yyyy HH:mm:ss Z",
        "d MMM yyyy HH:mm:ss Z",
        "EEE, d MMM yyyy HH:mm Z"
    ]

    static func parseDate(_ value: String) -> Date? {
        // Strip trailing comments like "(PDT)".
        var text = value
        if let paren = text.firstIndex(of: "(") {
            text = String(text[..<paren])
        }
        text = text.trimmingCharacters(in: .whitespaces)

        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        for format in dateFormats {
            formatter.dateFormat = format
            if let date = formatter.date(from: text) { return date }
        }
        return nil
    }

    // MARK: - Body / MIME

    private static func extractBody(_ data: Data,
                                    contentType: String,
                                    transferEncoding: String,
                                    depth: Int) -> String {
        guard depth < 5 else { return "" }
        let typeLower = contentType.lowercased()

        if typeLower.contains("multipart/"), let boundary = boundary(from: contentType) {
            let parts = splitMultipart(data, boundary: boundary)
            var htmlFallback = ""
            for part in parts {
                let (headers, partBody) = splitPart(part)
                let partType = headers["content-type"] ?? "text/plain"
                let partEnc = headers["content-transfer-encoding"] ?? ""
                let text = extractBody(partBody, contentType: partType,
                                       transferEncoding: partEnc, depth: depth + 1)
                if partType.lowercased().contains("text/plain"), !text.isEmpty {
                    return text
                }
                if htmlFallback.isEmpty, !text.isEmpty { htmlFallback = text }
            }
            return htmlFallback
        }

        // Leaf part: decode transfer encoding, then charset.
        let decoded: Data
        switch transferEncoding.lowercased().trimmingCharacters(in: .whitespaces) {
        case "base64":
            let compact = String(data: data, encoding: .ascii)?
                .components(separatedBy: .whitespacesAndNewlines).joined() ?? ""
            decoded = Data(base64Encoded: compact) ?? Data()
        case "quoted-printable":
            let text = String(data: data, encoding: .ascii) ?? ""
            decoded = decodeQuotedPrintable(text) ?? Data()
        default:
            decoded = data
        }

        var text = String(data: decoded, encoding: .utf8)
            ?? String(data: decoded, encoding: .isoLatin1) ?? ""

        if typeLower.contains("text/html") {
            text = stripHTML(text)
        }
        return text.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    private static func boundary(from contentType: String) -> String? {
        // boundary="xyz" or boundary=xyz
        let pattern = "boundary=\"?([^\";]+)\"?"
        guard let regex = try? NSRegularExpression(pattern: pattern, options: .caseInsensitive),
              let match = regex.firstMatch(in: contentType, range: NSRange(contentType.startIndex..., in: contentType)),
              let range = Range(match.range(at: 1), in: contentType) else { return nil }
        return String(contentType[range])
    }

    private static func splitMultipart(_ data: Data, boundary: String) -> [Data] {
        let marker = Data("--\(boundary)".utf8)
        var parts: [Data] = []
        var cursor = data.startIndex
        var partStart: Data.Index?

        while let range = data.range(of: marker, in: cursor..<data.endIndex) {
            if let start = partStart {
                parts.append(data.subdata(in: start..<range.lowerBound))
            }
            partStart = range.upperBound
            cursor = range.upperBound
            if cursor >= data.endIndex { break }
        }
        return parts
    }

    private static func splitPart(_ data: Data) -> ([String: String], Data) {
        let crlfSep = Data("\r\n\r\n".utf8)
        let lfSep = Data("\n\n".utf8)
        guard let r = data.range(of: crlfSep) ?? data.range(of: lfSep) else {
            return ([:], data)
        }
        let headerText = String(data: data.subdata(in: data.startIndex..<r.lowerBound), encoding: .utf8)
            ?? String(data: data.subdata(in: data.startIndex..<r.lowerBound), encoding: .isoLatin1) ?? ""
        let body = data.subdata(in: r.upperBound..<data.endIndex)
        return (unfoldHeaders(headerText), body)
    }

    static func decodeQuotedPrintable(_ text: String) -> Data? {
        var output = Data()
        var index = text.startIndex

        while index < text.endIndex {
            let char = text[index]
            if char == "=" {
                let next = text.index(after: index)
                // Soft line break: "=\n" or "=\r\n"
                if next < text.endIndex, text[next] == "\n" {
                    index = text.index(after: next); continue
                }
                if next < text.endIndex, text[next] == "\r" {
                    let afterCR = text.index(after: next)
                    if afterCR < text.endIndex, text[afterCR] == "\n" {
                        index = text.index(after: afterCR); continue
                    }
                }
                // Hex byte: "=XX"
                if let end = text.index(index, offsetBy: 3, limitedBy: text.endIndex) {
                    let hex = String(text[text.index(after: index)..<end])
                    if let byte = UInt8(hex, radix: 16) {
                        output.append(byte)
                        index = end
                        continue
                    }
                }
                output.append(UInt8(ascii: "="))
                index = text.index(after: index)
            } else {
                output.append(contentsOf: Array(String(char).utf8))
                index = text.index(after: index)
            }
        }
        return output
    }

    private static func stripHTML(_ html: String) -> String {
        html
            .replacingOccurrences(of: "<style[^>]*>[\\s\\S]*?</style>", with: " ", options: .regularExpression)
            .replacingOccurrences(of: "<script[^>]*>[\\s\\S]*?</script>", with: " ", options: .regularExpression)
            .replacingOccurrences(of: "<[^>]+>", with: " ", options: .regularExpression)
            .replacingOccurrences(of: "&nbsp;", with: " ")
            .replacingOccurrences(of: "&amp;", with: "&")
            .replacingOccurrences(of: "\\s+", with: " ", options: .regularExpression)
            .trimmingCharacters(in: .whitespacesAndNewlines)
    }
}

File: JobTracker/Services/MboxParser.cs
// Streaming parser for Google Takeout mbox archives. Reads the file in
// chunks (never loads the whole archive into memory), splits on mbox
// "From " separators, decodes MIME (RFC 2047 headers, quoted-printable /
// base64 bodies, multipart), and yields FetchedMessage values compatible
// with the rest of the pipeline.
//
// Takeout specifics used here:
// - Each message starts with "From <decimal-gm-thrid>@xxx <date>"
// - "X-GM-THRID" (decimal) converts to the hex threadId the Gmail API uses
// - "X-Gmail-Labels" lets us skip Spam / Trash outright

using System.Globalization;
using System.IO;
using System.Runtime.InteropServices;
using System.Text;
using System.Text.RegularExpressions;

namespace JobTracker.Services;

public sealed class MboxParserException(string message) : Exception(message);

public sealed class MboxParseSummary
{
    public int TotalMessages { get; set; }
    public int SkippedSpamTrash { get; set; }
    public int Candidates { get; set; }
}

public static partial class MboxParser
{
    private static readonly UTF8Encoding StrictUtf8 = new(encoderShouldEmitUTF8Identifier: false, throwOnInvalidBytes: true);

    /// Streams the mbox at `path`, returning only messages that pass the
    /// job-related prefilter (and are not Spam/Trash). `progress` receives
    /// a 0...1 fraction of file bytes consumed.
    public static (List<FetchedMessage> Messages, MboxParseSummary Summary) CollectJobCandidates(
        string path, Action<double>? progress = null)
    {
        if (!File.Exists(path)) throw new MboxParserException($"Cannot open mbox file: {path}");

        using var stream = File.OpenRead(path);
        var fileSize = stream.Length;
        const int chunkSize = 4 * 1024 * 1024;
        var separator = "\nFrom "u8.ToArray();

        var buffer = new List<byte>(chunkSize * 2);
        long bytesConsumed = 0;
        var summary = new MboxParseSummary();
        var results = new List<FetchedMessage>();

        void HandleMessage(byte[] raw)
        {
            summary.TotalMessages++;
            var parsed = ParseMessage(raw);
            if (parsed is null) return;
            if (parsed.Labels.Contains("Spam") || parsed.Labels.Contains("Trash"))
            {
                summary.SkippedSpamTrash++;
                return;
            }
            if (!HeuristicPrefilter.IsLikelyJobRelated(parsed.Fetched.Sender, parsed.Fetched.Subject,
                    parsed.Fetched.Body, parsed.Fetched.IsOutgoing))
            {
                return;
            }
            summary.Candidates++;
            results.Add(parsed.Fetched);
        }

        var readBuffer = new byte[chunkSize];
        int bytesRead;
        while ((bytesRead = stream.Read(readBuffer, 0, chunkSize)) > 0)
        {
            buffer.AddRange(bytesRead == chunkSize ? readBuffer : readBuffer[..bytesRead]);
            bytesConsumed += bytesRead;
            if (fileSize > 0) progress?.Invoke((double)bytesConsumed / fileSize);

            // Extract complete messages; keep the trailing partial one buffered.
            var searchStart = 0;
            while (true)
            {
                var bufSpan = CollectionsMarshal.AsSpan(buffer);
                if (searchStart >= bufSpan.Length) break;
                var relIdx = bufSpan[searchStart..].IndexOf(separator);
                if (relIdx < 0) break;
                var idx = searchStart + relIdx;

                // Validate this is a real mbox boundary: "From <digits>@xxx"
                if (!IsTakeoutBoundary(bufSpan, idx + separator.Length))
                {
                    searchStart = idx + separator.Length;
                    continue;
                }

                var messageBytes = buffer.GetRange(0, idx).ToArray();
                if (messageBytes.Length > 0) HandleMessage(messageBytes);
                // +1 keeps the "From " (drops only the "\n").
                buffer.RemoveRange(0, idx + 1);
                searchStart = 0;
            }
        }

        // Final message.
        if (buffer.Count > 0) HandleMessage(buffer.ToArray());

        return (results, summary);
    }

    /// Checks that the bytes after "\nFrom " look like "<digits>@xxx ".
    private static bool IsTakeoutBoundary(ReadOnlySpan<byte> data, int index)
    {
        var sawDigit = false;
        while (index < data.Length && data[index] is >= (byte)'0' and <= (byte)'9')
        {
            sawDigit = true;
            index++;
        }
        return sawDigit && index < data.Length && data[index] == (byte)'@';
    }

    // MARK: - Single message

    private sealed record ParsedMessage(FetchedMessage Fetched, HashSet<string> Labels);

    private static ParsedMessage? ParseMessage(byte[] raw)
    {
        var (headerBytes, bodyBytes) = SplitHeaderBody(raw);

        var headerText = DecodeUtf8OrLatin1(headerBytes);
        var headers = UnfoldHeaders(headerText);
        string Header(string name) => headers.GetValueOrDefault(name.ToLowerInvariant(), "");

        var labels = Header("X-Gmail-Labels").Split(',').Select(l => l.Trim())
            .Where(l => l.Length > 0).ToHashSet();

        var threadIdHex = "";
        if (ulong.TryParse(Header("X-GM-THRID"), NumberStyles.None, CultureInfo.InvariantCulture, out var decimalId))
        {
            threadIdHex = decimalId.ToString("x", CultureInfo.InvariantCulture);
        }

        var messageId = Header("Message-ID").Trim('<', '>', ' ', '\t');

        var subject = DecodeRfc2047(Header("Subject"));
        var sender = DecodeRfc2047(Header("From"));
        var date = ParseDate(Header("Date")) ?? DateTimeOffset.MinValue;

        var body = ExtractBody(bodyBytes, Header("Content-Type"), Header("Content-Transfer-Encoding"), 0);
        var cleanBody = Truncate(body, 8000);

        var fetched = new FetchedMessage(
            Id: messageId.Length == 0 ? $"mbox-{threadIdHex}-{raw.Length}" : messageId,
            ThreadId: threadIdHex,
            Sender: sender,
            Subject: subject,
            Snippet: Truncate(cleanBody.Replace('\n', ' '), 160),
            Body: cleanBody,
            Date: date,
            IsOutgoing: labels.Contains("Sent"));

        return new ParsedMessage(fetched, labels);
    }

    // MARK: - Headers

    private static (byte[] Header, byte[] Body) SplitHeaderBody(ReadOnlySpan<byte> raw)
    {
        var crlf = "\r\n\r\n"u8;
        var lf = "\n\n"u8;
        var idx = raw.IndexOf(crlf);
        var sepLen = 4;
        if (idx < 0) { idx = raw.IndexOf(lf); sepLen = 2; }
        if (idx < 0) return (raw.ToArray(), []);
        return (raw[..idx].ToArray(), raw[(idx + sepLen)..].ToArray());
    }

    /// Unfolds continuation lines and returns a [lowercased-name: value] map
    /// (first occurrence wins).
    private static Dictionary<string, string> UnfoldHeaders(string text)
    {
        var result = new Dictionary<string, string>();
        string? currentName = null;
        var currentValue = new StringBuilder();

        void Commit()
        {
            if (currentName is not null && !result.ContainsKey(currentName))
            {
                result[currentName] = currentValue.ToString().Trim();
            }
        }

        foreach (var line in text.Replace("\r\n", "\n").Split('\n'))
        {
            if (line.StartsWith(' ') || line.StartsWith('\t'))
            {
                currentValue.Append(' ').Append(line.Trim());
            }
            else
            {
                var colon = line.IndexOf(':');
                if (colon >= 0)
                {
                    Commit();
                    currentName = line[..colon].ToLowerInvariant();
                    currentValue.Clear();
                    currentValue.Append(line[(colon + 1)..]);
                }
            }
        }
        Commit();
        return result;
    }

    [GeneratedRegex(@"=\?([^?]+)\?([BbQq])\?([^?]*)\?=")]
    private static partial Regex Rfc2047Pattern();

    /// Decodes RFC 2047 encoded words: =?charset?B|Q?data?=
    public static string DecodeRfc2047(string value)
    {
        if (!value.Contains("=?")) return value;

        var result = value;
        var iterations = 0;
        while (iterations < 20)
        {
            var match = Rfc2047Pattern().Match(result);
            if (!match.Success) break;
            iterations++;

            var charset = match.Groups[1].Value.ToLowerInvariant();
            var encoding = match.Groups[2].Value.ToUpperInvariant();
            var payload = match.Groups[3].Value;

            byte[]? decodedData;
            if (encoding == "B")
            {
                try { decodedData = Convert.FromBase64String(payload); }
                catch (FormatException) { decodedData = null; }
            }
            else
            {
                decodedData = DecodeQuotedPrintable(payload.Replace('_', ' '));
            }

            var text = decodedData is null ? ""
                : charset.Contains("8859") ? Encoding.Latin1.GetString(decodedData)
                : DecodeUtf8OrLatin1(decodedData);

            result = string.Concat(result.AsSpan(0, match.Index), text, result.AsSpan(match.Index + match.Length));
        }
        return result;
    }

    // MARK: - Date

    private static readonly string[] DateFormats =
    [
        "ddd, d MMM yyyy HH:mm:ss zzz",
        "d MMM yyyy HH:mm:ss zzz",
        "ddd, d MMM yyyy HH:mm zzz",
    ];

    [GeneratedRegex(@"([+-]\d{2})(\d{2})$")]
    private static partial Regex TrailingOffsetRegex();

    public static DateTimeOffset? ParseDate(string value)
    {
        // Strip trailing comments like "(PDT)".
        var text = value;
        var paren = text.IndexOf('(');
        if (paren >= 0) text = text[..paren];
        text = text.Trim();

        // Normalize "-0700" → "-07:00" so the "zzz" custom specifier parses it.
        var offsetMatch = TrailingOffsetRegex().Match(text);
        if (offsetMatch.Success)
        {
            text = string.Concat(text.AsSpan(0, offsetMatch.Index),
                offsetMatch.Groups[1].Value, ":", offsetMatch.Groups[2].Value);
        }

        foreach (var format in DateFormats)
        {
            if (DateTimeOffset.TryParseExact(text, format, CultureInfo.InvariantCulture,
                    DateTimeStyles.AllowWhiteSpaces, out var date))
            {
                return date;
            }
        }
        return null;
    }

    // MARK: - Body / MIME

    private static string ExtractBody(byte[] data, string contentType, string transferEncoding, int depth)
    {
        if (depth >= 5) return "";
        var typeLower = contentType.ToLowerInvariant();

        if (typeLower.Contains("multipart/"))
        {
            var boundary = Boundary(contentType);
            if (boundary is not null)
            {
                var parts = SplitMultipart(data, boundary);
                var htmlFallback = "";
                foreach (var part in parts)
                {
                    var (headers, partBody) = SplitPart(part);
                    var partType = headers.GetValueOrDefault("content-type", "text/plain");
                    var partEnc = headers.GetValueOrDefault("content-transfer-encoding", "");
                    var text = ExtractBody(partBody, partType, partEnc, depth + 1);
                    if (partType.ToLowerInvariant().Contains("text/plain") && text.Length > 0) return text;
                    if (htmlFallback.Length == 0 && text.Length > 0) htmlFallback = text;
                }
                return htmlFallback;
            }
        }

        // Leaf part: decode transfer encoding, then charset.
        byte[] decoded;
        switch (transferEncoding.Trim().ToLowerInvariant())
        {
            case "base64":
                var compact = new string(Encoding.ASCII.GetString(data)
                    .Where(c => !char.IsWhiteSpace(c)).ToArray());
                try { decoded = Convert.FromBase64String(compact); }
                catch (FormatException) { decoded = []; }
                break;
            case "quoted-printable":
                decoded = DecodeQuotedPrintable(Encoding.ASCII.GetString(data)) ?? [];
                break;
            default:
                decoded = data;
                break;
        }

        var text2 = DecodeUtf8OrLatin1(decoded);
        if (typeLower.Contains("text/html")) text2 = StripHtml(text2);
        return text2.Trim();
    }

    [GeneratedRegex("boundary=\"?([^\";]+)\"?", RegexOptions.IgnoreCase)]
    private static partial Regex BoundaryRegex();

    private static string? Boundary(string contentType)
    {
        var match = BoundaryRegex().Match(contentType);
        return match.Success ? match.Groups[1].Value : null;
    }

    private static List<byte[]> SplitMultipart(byte[] data, string boundary)
    {
        var marker = Encoding.UTF8.GetBytes($"--{boundary}");
        var parts = new List<byte[]>();
        var cursor = 0;
        int? partStart = null;

        while (true)
        {
            var relIdx = ((ReadOnlySpan<byte>)data)[cursor..].IndexOf(marker);
            if (relIdx < 0) break;
            var idx = cursor + relIdx;
            if (partStart is int start)
            {
                parts.Add(data[start..idx]);
            }
            partStart = idx + marker.Length;
            cursor = idx + marker.Length;
            if (cursor >= data.Length) break;
        }
        return parts;
    }

    private static (Dictionary<string, string> Headers, byte[] Body) SplitPart(byte[] data)
    {
        var (headerBytes, body) = SplitHeaderBody(data);
        if (body.Length == 0 && headerBytes.Length == data.Length)
        {
            // No blank-line separator found: SplitHeaderBody returned the
            // whole thing as "header" with an empty body — treat as body-only.
            return ([], data);
        }
        return (UnfoldHeaders(DecodeUtf8OrLatin1(headerBytes)), body);
    }

    public static byte[]? DecodeQuotedPrintable(string text)
    {
        var output = new List<byte>(text.Length);
        var index = 0;

        while (index < text.Length)
        {
            var c = text[index];
            if (c == '=')
            {
                var next = index + 1;
                // Soft line break: "=\n" or "=\r\n"
                if (next < text.Length && text[next] == '\n') { index = next + 1; continue; }
                if (next < text.Length && text[next] == '\r')
                {
                    var afterCr = next + 1;
                    if (afterCr < text.Length && text[afterCr] == '\n') { index = afterCr + 1; continue; }
                }
                // Hex byte: "=XX"
                if (index + 3 <= text.Length &&
                    byte.TryParse(text.AsSpan(index + 1, 2), NumberStyles.AllowHexSpecifier,
                        CultureInfo.InvariantCulture, out var b))
                {
                    output.Add(b);
                    index += 3;
                    continue;
                }
                output.Add((byte)'=');
                index++;
            }
            else
            {
                output.AddRange(Encoding.UTF8.GetBytes(c.ToString()));
                index++;
            }
        }
        return [.. output];
    }

    [GeneratedRegex(@"<style[^>]*>[\s\S]*?</style>", RegexOptions.IgnoreCase)]
    private static partial Regex StyleTagRegex();
    [GeneratedRegex(@"<script[^>]*>[\s\S]*?</script>", RegexOptions.IgnoreCase)]
    private static partial Regex ScriptTagRegex();
    [GeneratedRegex(@"<[^>]+>")]
    private static partial Regex TagRegex();
    [GeneratedRegex(@"\s+")]
    private static partial Regex WhitespaceRegex();

    private static string StripHtml(string html)
    {
        var result = StyleTagRegex().Replace(html, " ");
        result = ScriptTagRegex().Replace(result, " ");
        result = TagRegex().Replace(result, " ");
        result = result.Replace("&nbsp;", " ").Replace("&amp;", "&");
        result = WhitespaceRegex().Replace(result, " ");
        return result.Trim();
    }

    // MARK: - Shared helpers

    private static string DecodeUtf8OrLatin1(byte[] bytes)
    {
        try { return StrictUtf8.GetString(bytes); }
        catch (DecoderFallbackException) { return Encoding.Latin1.GetString(bytes); }
    }

    private static string Truncate(string value, int maxLength) =>
        value.Length <= maxLength ? value : value[..maxLength];
}
